src/main/java/com/javampire/openscad/color/OpenSCADColorProvider.java

```diff
@@ -2,31 +2,55 @@
 
 import com.intellij.openapi.editor.ElementColorProvider;
 import com.intellij.psi.PsiElement;
+import com.intellij.psi.PsiFileFactory;
 import com.intellij.psi.util.PsiTreeUtil;
 import com.intellij.xml.util.ColorMap;
+import com.javampire.openscad.OpenSCADFileType;
 import com.javampire.openscad.psi.OpenSCADTypes;
-import com.javampire.openscad.psi.impl.OpenSCADBuiltinOpImpl;
+import com.javampire.openscad.psi.impl.OpenSCADBuiltinObjImpl;
+
 import org.jetbrains.annotations.NotNull;
 import org.jetbrains.annotations.Nullable;
 
-import java.awt.*;
+import java.awt.Color;
 
 class OpenSCADColorProvider implements ElementColorProvider {
 
     @Nullable
     @Override
     public Color getColorFrom(@NotNull PsiElement element) {
-        if (element.getNode().getElementType() == OpenSCADTypes.STRING_LITERAL) {
-            final PsiElement colorParent = PsiTreeUtil.getParentOfType(element, OpenSCADBuiltinOpImpl.class);
-            if (colorParent != null && colorParent.getText().startsWith("color")) {
-                final String colorDef = element.getText();
-                return ColorMap.getColor(colorDef.substring(1, colorDef.length() - 1));
-            }
+        final PsiElement fullColorBlock = getOpenSCADBuiltinObj(element);
+        if (fullColorBlock != null) {
+            final String colorDef = element.getText();
+            return ColorMap.getColor(colorDef.substring(1, colorDef.length() - 1));
         }
+
         return null;
     }
 
     @Override
     public void setColorTo(@NotNull PsiElement element, @NotNull Color color) {
+        final PsiElement fullColorBlock = getOpenSCADBuiltinObj(element);
+        if (fullColorBlock != null) {
+            final String newText = createColorHex(color) + fullColorBlock.getChildren()[1].getText();
+
+            final PsiElement newElement = PsiFileFactory.getInstance(element.getProject())
+                    .createFileFromText("tmp.scad", OpenSCADFileType.INSTANCE, newText).getFirstChild();
```

Comment:
Can this be the source of the bug ? Might be that the file is not overwritten. Anyway, it does not seems like a good solution to generate a file, it is generally slow.

Reply:
https://www.jetbrains.org/intellij/sdk/docs/basics/architectural_overview/modifying_psi.html#creating-the-new-psi